### PR TITLE
testdrive: rewrite protobuf tests

### PR DESCRIPTION
**Pull request by benesch**
@cjubb39 this was inspired by your comment here: https://github.com/MaterializeInc/materialize/pull/9019#pullrequestreview-803251518. I found it really hard to discover precisely what we were and weren't testing, so I went for a rather major refactor.

----

This commit is a major overhaul of our protobuf testing infrastructure.
There are two big highlights:

  * The unit tests in the `interchange` crate are replaced with
    equivalent tests in testdrive script (when those tests don't already
    exist). The unit tests were tightly coupled to the current design of
    the `interchange::protobuf` module, which impeded refactors like the
    one in #9019. They're also just quite a bit more verbose than the
    equivalent testdrive tests.

  * Testdrive has learned to compile .proto files on the fly. This means
    that protobuf test scripts can include the desired messages in the
    same .td file that manipulates sources and Kafka topics based on
    those messages. This makes the tests vastly more readable, in my
    opinion.

I also took the opportunity to split the monolithic `protobuf.td` file
into multiple pieces, each of which I hope is much easier to follow. I
was quite aggressive about pruning the scope of each test. When
possible, I only tested each feature in one place.

So, while it may look like this commit decreases test coverage in terms
of lines changed, I think it actually *increases* test coverage (modulo
any mistakes I've made). For example, we now actually test custom
defaults in proto2 messages.

### Motivation

   * This PR refactors existing code.

### Checklist

- [x] This PR has adequate test coverage / QA involvement has been duly considered.
- [x] This PR adds a release note for any
  [user-facing behavior changes](https://github.com/MaterializeInc/materialize/blob/main/doc/user/content/release-notes.md#what-changes-require-a-release-note).

### CODE REVIEW

File: src/protoc/Cargo.toml [hunk 1]
```diff
@@ -12,9 +12,8 @@ path = "bin/protoc.rs"
 [dependencies]
 anyhow = "1.0.47"
 ore = { path = "../ore" }
-protobuf = "3.0.0-alpha.2"
-protobuf-codegen = "3.0.0-alpha.2"
-protobuf-codegen-pure = "3.0.0-alpha.2"
-protobuf-parse = "3.0.0-alpha.2"
+protobuf = { git = "https://github.com/MaterializeInc/rust-protobuf.git" }
```
Comment (cjubb39) [inline src/protoc/Cargo.toml:15]: this commit is the only difference? https://github.com/MaterializeInc/rust-protobuf/commit/063964a8ab6fee98a9622a0c0a72a9765411025a

or can you summarize why we're forking?  It'll add maintenance cost for us so I want to make sure I understand

File: src/testdrive/src/action.rs [hunk 2]
```diff
@@ -350,19 +350,6 @@ pub async fn build(cmds: Vec<PosCommand>, state: &State) -> Result<Vec<PosAction
         "testdrive.temp-dir".into(),
         state.temp_path.display().to_string(),
     );
-    {
```
Comment (cjubb39) [inline src/testdrive/src/action.rs:353]: I love this deletion lol

File: src/testdrive/src/action/kafka/ingest.rs [hunk 3]
```diff
@@ -224,9 +191,15 @@ pub fn build_ingest(mut cmd: BuiltinCommand) -> Result<IngestAction, String> {
         None => None,
     };
     let timestamp = cmd.args.opt_parse("timestamp")?;
-    let publish = cmd.args.opt_bool("publish")?.unwrap_or(false);
     cmd.args.done()?;
 
+    if publish
```
Comment (cjubb39) [inline src/testdrive/src/action/kafka/ingest.rs:196]: why this change?

fwiw my take is that the `publish` arg here makes it harder to reconstruct what a user would actually need to do in order to properly use a schema -- but I'm curious why you decided to remove it for protos here?  Just because it's confusing that the behavior differs from confluent tools in that it doesn't follow references?

File: test/testdrive/protobuf-import.td [hunk 4]
```diff
@@ -0,0 +1,105 @@
...
+  int32 nanos = 2;
+}
+
+$ schema-registry-publish subject=importee.proto schema-type=protobuf
+\${importee-schema}
+
+$ schema-registry-publish subject=testdrive-import-csr-${testdrive.seed}-value schema-type=protobuf references=importee.proto,google/protobuf/timestamp.proto
+\${importer-schema}
+
+# TODO(benesch): There is a big bug here! In the Confluent toolchain, these
```
Comment (cjubb39) [inline test/testdrive/protobuf-import.td:88]: Can you link a task? since this is a pretty big one we should fix sooner rather than later!

File: test/testdrive/protobuf-repeated.td [hunk 5]
```diff
@@ -0,0 +1,56 @@
...
+
+$ kafka-ingest topic=repeated format=protobuf descriptor-file=repeated.pb message=.Repeated
+{"bool": [true, false], "enum": ["ENUM1", "ENUM0"], "message": [{"m1": 4, "m2": 2}, {"m1": 2, "m2": 4}]}
+{}
+
+> CREATE MATERIALIZED SOURCE repeated FROM
+  KAFKA BROKER '${testdrive.kafka-addr}' TOPIC 'testdrive-repeated-${testdrive.seed}'
+  FORMAT PROTOBUF MESSAGE '.Repeated' USING SCHEMA FILE '${testdrive.temp-dir}/repeated.pb'
+
+# TODO(benesch): all of these columns should be non-nullable.
```
Comment (cjubb39) [inline test/testdrive/protobuf-repeated.td:43]: This and similar errors in other files are still planned to be fixed by your #9019 yeah?

File: test/testdrive/protobuf-required.td [hunk 6]
```diff
@@ -0,0 +1,23 @@
...
+# Test that required Protobuf fields are unsupported.
+
+$ file-append path=required.proto
+syntax = "proto2";
+
+message Required {
+  required int32 f = 1;
+}
+
+$ protobuf-compile-descriptors inputs=required.proto output=required.pb
```
Comment (cjubb39) [inline test/testdrive/protobuf-required.td:19]: nice! I think these files, especially because they're so concise now, are good examples for anyone actually trying to use these types of sources.

e.g. the format of the input FILE below was something I was confused by for a little bit, partly because of the magic in testdrive.  This "explicitly compile the proto and then use it in the `CREATE SOURCE` statement" flow is much more clear!

File: test/testdrive/protobuf-import.td [hunk 7]
```diff
@@ -0,0 +1,105 @@
...
+
+message Importee2 {
+    int32 i = 1;
+}
+
+
+$ set importer-schema
+syntax = "proto3";
+
+import "google/protobuf/timestamp.proto";
```
Comment (cjubb39) [inline test/testdrive/protobuf-import.td:27]: What do you think of moving this import to one of the messages in `importee-schema`? I think it could be good to exercise chains of references that need to be imported from CSR -- so we're confident that we're not just importing top level references

File: test/testdrive/protobuf-basic.td [hunk 8]
```diff
@@ -0,0 +1,68 @@
...
+    double double = 5;
+    bytes bytes = 6;
+    string string = 7;
+    Enum enum = 8;
+    Message message = 9;
+}
+
+$ protobuf-compile-descriptors inputs=basic.proto output=basic.pb
+
+$ kafka-create-topic topic=basic partitions=1
```
Comment (cjubb39) [inline test/testdrive/protobuf-basic.td:39]: Looks like one thing that was dropped when splitting out / rewriting files was multi-partition topics.

If this was intentional, can you drop a short note explaining why it's okay?  It's not immediately obvious to me

**Review by cjubb39 — COMMENTED** [review]
Looks good to me!  I think testing will be much easier to follow with this change -- and build confidence that #9019 / similar in the future are correct

A couple questions inline.  Some are just to make sure I understand some whys  but there are two that I'd like to see addressed before merging: 1) schema registry having more recursive testing and 2) multi-partition test seemingly missing

File: src/protoc/Cargo.toml (hunk 1 above)
Comment (benesch) [reply src/protoc/Cargo.toml:15]: Ah yeah, it's just that PR (https://github.com/stepancheg/rust-protobuf/pull/581). I dropped a note in deny.toml with a link to the PR.

File: test/testdrive/protobuf-repeated.td (hunk 5 above)
Comment (benesch) [reply test/testdrive/protobuf-repeated.td:50]: Yep exactly!

File: src/testdrive/src/action/kafka/ingest.rs (hunk 3 above)
Comment (benesch) [reply src/testdrive/src/action/kafka/ingest.rs:198]: Yeah, it's just super hard to support properly for protos. `kakfa-ingest` with proto-formatted messages now takes a descriptor file and the name of a message in that descriptor file. Unfortunately that's not enough to publish the schema to the CSR: you need the input`.proto` file, not the compiled file descriptor set. But actually even _that's_ not quite right, because AFAICT the Confluent stack will reconstruct the definition of the _one_ proto message that will actually appear in the topic, rather than publishing the entire `.proto` file. The Rust protobuf library just doesn't support taking a `MessageDescriptor` and rendering it as a `.proto` file (while the Java library does). So seemed way easier to just mandate that we manually publish Protobuf schemas to the registry.

File: test/testdrive/protobuf-basic.td (hunk 8 above)
Comment (benesch) [reply test/testdrive/protobuf-basic.td:46]: Ah yeah, just didn't seem like a necessary thing to test. There's nothing special about protobuf that would interact weirdly with multi-partition topics, and we've got plenty of multi-partition topics elsewhere.

File: test/testdrive/protobuf-import.td (hunk 4 above)
Comment (benesch) [reply test/testdrive/protobuf-import.td:100]: Yep, done!

File: test/testdrive/protobuf-import.td (hunk 7 above)
Comment (benesch) [reply test/testdrive/protobuf-import.td]: Ooh good idea. I moved the timestamp import as you suggest, and replaced it with a useless import of an empty file, to test that we continue to handle schemas with multiple references correctly.

File: test/testdrive/protobuf-basic.td (hunk 8 above)
Comment (cjubb39) [reply test/testdrive/protobuf-basic.td:46]: makes sense!

File: test/testdrive/protobuf-defaults.td [hunk 9]
```diff
@@ -0,0 +1,47 @@
...
+
+$ kafka-ingest topic=defaults format=protobuf descriptor-file=defaults.pb message=.Defaults
+{}
+
+> CREATE MATERIALIZED SOURCE defaults FROM
+  KAFKA BROKER '${testdrive.kafka-addr}' TOPIC 'testdrive-defaults-${testdrive.seed}'
+  FORMAT PROTOBUF MESSAGE '.Defaults' USING SCHEMA FILE '${testdrive.temp-dir}/defaults.pb'
+
+# TODO(benesch): The fact that ENUM0 is returned below is incorrect. It should
+# be ENUM1.
```
Comment (cjubb39) [inline test/testdrive/protobuf-defaults.td:43]: Is this something you have a fix for too?  I'm not sure if an issue is too heavyweight but I'm worried we might forget this if the fix doesn't come soon

**Review by cjubb39 — APPROVED** [review]
nice! let's do it!

File: test/testdrive/protobuf-defaults.td (hunk 9 above)
Comment (benesch) [reply test/testdrive/protobuf-defaults.td:43]: 💯 I do! Normally I would file an issue but I've already got the fix prepared. Coming your way soon!
